app/background.py

```diff
@@ -46,22 +46,39 @@ def put(self, data):
 
 
 def background(inputQueue, outputQueue):
-  block = True
   pid = os.getpid()
   signalNumber = signal.SIGUSR1
+  def redrawProgram(program, callerSemaphore):
+    """
+    Sends a SIGUSR1 signal to the current program and draws its screen.
+
+    Args:
+      program (CiProgram): an instance of the CiProgram object.
+
+    Returns:
+      None.
+    """
+    program.render()
+    outputQueue.put((app.render.frame.grabFrame(), callerSemaphore))
+    os.kill(pid, signalNumber)
+
+  block = True
   while True:
     try:
       try:
-        program, message = inputQueue.get(block)
+        program, message, callerSemaphore = inputQueue.get(block)
         #profile = app.profile.beginPythonProfile()
         if message == 'quit':
           app.log.info('bg received quit message')
           return
+        elif message == 'popup':
```

Comment:
Can the pop-up simply be a top level modal window in the ci_program.py zOrder, so that background.py wouldn't need to know anything about it?

Reply:
What do you mean by this? I'm making it so that the FileStats object contacts the background thread, and the background thread contacts the main thread. Right now background.py doesn't really know anything about the pop-up and it's just redirecting it to the main thread.